cvat/apps/engine/ddln/tasks/vls_lines/models.py

```diff
@@ -0,0 +1,139 @@
+import math
+from itertools import combinations
+from typing import Optional
+
+from cvat.apps.engine.ddln.geometry import Line, Point, get_angle_between, get_counterclockwise_angle
+
+ANGLE_THRESHOLD = 0.05
+
+
+class Runway:
+    def __init__(
+        self,
+        id: str,
+        left_line: Optional[Line],
+        right_line: Optional[Line],
+        center_line: Optional[Line],
+        start_line: Optional[Line],
+        end_line: Optional[Line],
+        designator_line: Optional[Line],
+    ):
+        self.id = id
+        self.left_line = left_line
+        self.right_line = right_line
+        self.center_line = center_line
+        self.start_line = start_line
+        self.end_line = end_line
+        self.designator_line = designator_line
+        self.lon_vanishing_point = None
+        self.lat_vanishing_point = None
+
+    @property
+    def track_id(self):
+        return self.id
+
+    def calculate_vanishing_points(self, reporter):
+        lon_lines = [self.left_line, self.center_line, self.right_line]
+        self.lon_vanishing_point = self._calculate_vanishing_point(lon_lines, reporter, is_lon=True)
+        lat_lines = [self.start_line, self.designator_line, self.end_line]
+        if all(line is not None for line in lat_lines):
+            self.lat_vanishing_point = self._calculate_vanishing_point(lat_lines, reporter, is_lon=False)
+
+    def fix_order(self, reporter):
+        if self.lon_vanishing_point:
+            self._check_lon_order(reporter)
+        else:
+            self._check_parallel_lon_order(reporter)
+        self._check_lat_order(reporter)
+
+    def apply_visibility(
+        self,
+        left_visible,
+        right_visible,
+        center_visible,
+        start_visible,
+        end_visible,
+        designator_visible
+    ):
+        if not left_visible:
+            self.left_line = None
+        if not right_visible:
+            self.right_line = None
+        if not center_visible:
+            self.center_line = None
+        if not start_visible:
+            self.start_line = None
+        if not end_visible:
+            self.end_line = None
+        if not designator_visible:
+            self.designator_line = None
+
+    def _check_lon_order(self, reporter):
+        original = [self.left_line, self.center_line, self.right_line]
+        max_angle = 0
+        edges = None
+        for a, b in combinations(original, 2):
+            angle = get_lines_angle(a, b)
+            if angle > max_angle:
+                max_angle = angle
+                edges = [a, b]
+        new_center = next(filter(lambda line: line not in edges, original))
+        new_left, new_right = edges
+        if get_counterclockwise_angle(new_right.get_angle(), new_left.get_angle()) >= math.pi:
```

Comment:
This check validates that the annotator didn't mixed up 'left' and 'right' edge, it doesn't depend on the rotation of the image (it has the assumption that the angle between left and right edge is < math.pi / 2 though, so it might give false-positive).

It checks counterclockwise angle (rather than clockwise) because cvat y axis is opposite to output/drawing y axis.

![desc](https://user-images.githubusercontent.com/14096024/102613491-a2826180-4143-11eb-8dc3-622e4b63d93b.png)